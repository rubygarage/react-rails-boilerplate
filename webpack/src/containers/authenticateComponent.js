import React from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { getCurrentUser } from 'selectors/user';

function authenticateComponent(WrappedComponent) {
  const mapStateToProps = state => ({ currentUser: getCurrentUser(state) });
  return connect(mapStateToProps)((props) => {
    if (!props.currentUser.id) { return <Redirect to="/sign_in" />; }
    return <WrappedComponent {...props} />;
  });
}

export default authenticateComponent;
