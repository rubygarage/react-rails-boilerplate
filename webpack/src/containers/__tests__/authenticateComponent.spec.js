import React from 'react';
import authenticateComponent from '../authenticateComponent';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';

describe('authenticateComponent HoC', () => {
  const defaultProps = {
    match: {
      params: {
        id: '1',
      },
    },
  };
  const MockComponent = () => <div>Mock Component</div>;
  const WrappedComponent = authenticateComponent(MockComponent);

  describe('user is authenticated', () => {
    const store = configureStore()({
      entities: { users: { 1: { id: '1', username: 'name' } } },
      signin: { currentUser: { users: ['1'] } },
    });

    it('renders given component', () => {
      const authenticatedContainer = mount(<WrappedComponent {...defaultProps} store={store} />);
      expect(authenticatedContainer).toMatchSnapshot();
    });
  });

  describe('user is not authenticated', () => {
    const unauthenticatedStore = configureStore()({
      entities: { users: {} },
      signin: { currentUser: { users: [] } },
    });

    it('renders Redirect component if user is not authenticated', () => {
      const unauthenticatedContainer = mount((
        // keyLength={0} to stop snapshots from failing due to randomly generated keys
        <MemoryRouter keyLength={0}>
          <WrappedComponent
            {...defaultProps}
            store={unauthenticatedStore}
          />
        </MemoryRouter>
      ));
      expect(unauthenticatedContainer).toMatchSnapshot();
    });
  });
});
