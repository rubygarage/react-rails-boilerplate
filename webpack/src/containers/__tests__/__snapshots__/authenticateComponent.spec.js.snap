// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticateComponent HoC user is authenticated renders given component 1`] = `
<Connect(Component)
  match={
    Object {
      "params": Object {
        "id": "1",
      },
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Component
    currentUser={
      Object {
        "id": "1",
        "username": "name",
      }
    }
    dispatch={[Function]}
    match={
      Object {
        "params": Object {
          "id": "1",
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <MockComponent
      currentUser={
        Object {
          "id": "1",
          "username": "name",
        }
      }
      dispatch={[Function]}
      match={
        Object {
          "params": Object {
            "id": "1",
          },
        }
      }
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
      storeSubscription={
        Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        }
      }
    >
      <div>
        Mock Component
      </div>
    </MockComponent>
  </Component>
</Connect(Component)>
`;

exports[`authenticateComponent HoC user is not authenticated renders Redirect component if user is not authenticated 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "REPLACE",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/sign_in",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/sign_in",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(Component)
      match={
        Object {
          "params": Object {
            "id": "1",
          },
        }
      }
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Component
        currentUser={Object {}}
        dispatch={[Function]}
        match={
          Object {
            "params": Object {
              "id": "1",
            },
          }
        }
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
        }
      >
        <Redirect
          push={false}
          to="/sign_in"
        />
      </Component>
    </Connect(Component)>
  </Router>
</MemoryRouter>
`;
